### PR TITLE
ADBDEV-8057: Fix pg_rewind leaving replica on old timeline

### PR DESCRIPTION
**Pull request by silent-observer**
Remove bogus assertion in pg_rewind

It's entirely possible that the calculated "point of divergence" is
before the end of WAL on the target, because when we calculate the
point of divergence, we assume that the target's WAL extends to
infinity on its current timeline. It might be more clear to calculate
the end of WAL first, and if the target is a direct ancestor of the
source, consider the end of target WAL as the point of
divergence. However, it doesn't change the result: the target is a
direct ancestor of the source, and doesn't need to be rewound.

Add tests to cover the "no rewind required" cases, both when the
target's end of WAL is exactly at the point of divergence, which is
what the removed assertion assumed (case B in the test), and the case
where the assertion failed (case A).

Changes from original patch: test was made more general to prepare for more
detailed tests in the next commits. In particular, we pg_rewind from a
running source node, not from a stopped one, and the source node is passed as
an argument. This doesn't affect the test validity for this patch.

Discussion: https://www.postgresql.org/message-id/18575-1b9b3b60e2a480de@postgresql.org

---

Alternatively, perhaps just copying the history file would be enough?

Changes from original patch:
- Move sourceHistfile initialization to before rewind_parseTimeLineHistory,
  since rewind_parseTimeLineHistory actually modifies it.
- Don't copy timeline history file when dry_run is specified. Dry run
  shouldn't make changes that can potentially break the target server when
  applied incorrectly.
- Fix 107_empty_conf.pl test to preserve primary_conninfo setting.

---

Fix pg_rewind edge case when checkpoint is at switchpoint

In one specific edge case, when the source server makes a checkpoint exactly
after switching to a new timeline, switchpoint LSN exactly matches latest
checkpoint LSN. This can happen, for example, if the source has just performed
archive recovery. Then setting minRecoveryPoint to the checkpoint LSN in
pg_rewind is illegal and causes an error at startup, since minRecoveryPoint
actually means the minimum allowed end of WAL, and we want that to be after
the switch occurs, not at the end of the previous timeline.

As a fix, add +1 to the checkpoint LSN when setting minRecoveryPoint, which
now correctly handles this case. Also add a test scenario for the source node
that has performed recovery.

---

Note: Do not squash to preserve authorship

### CODE REVIEW

File: src/bin/pg_rewind/pg_rewind.c [hunk 1]
```diff
@@ -372,12 +375,39 @@ main(int argc, char **argv)
...
+				update_controlfile(datadir_target, &ControlFile_new, do_sync);
+
+				/*
+				 * Make the timeline history file available to the target. Avoids
+				 * having to fetch it immediately at startup, and ensures that it
+				 * has the same idea of what the timeline is as we did.
+				 */
+				open_target_file(sourceHistfileName, true);
+				write_target_range(sourceHistfile, 0, strlen(sourceHistfile));
+				close_target_file();
```
Comment (RekGRpth) [inline src/bin/pg_rewind/pg_rewind.c]: The original patch does this unconditionally.

**Comment by RekGRpth** [comment]
Add a description of the changes in the `010_target_is_ancestor.pl` test to the description of the second commit.

**Comment by RekGRpth** [comment]
Maybe we should separate the changes to the `010_target_is_ancestor.pl` file into a separate third commit with a detailed description? (like in https://github.com/arenadata/gpdb/pull/1189)

File: src/bin/pg_rewind/t/010_target_is_ancestor.pl [hunk 2]
```diff
@@ -55,25 +55,30 @@
 $node_2->init_from_backup($orig_node, 'backup_B', has_streaming => 1);
 $node_2->start;
 $node_2->promote;
-$node_2->safe_psql('postgres', "INSERT INTO test_tab VALUES ('in node_2')");
-$node_2->stop;
+$node_2->safe_psql('postgres', "INSERT INTO test_tab VALUES ('in node')");
```
Comment (KnightMurloc) [inline src/bin/pg_rewind/t/010_target_is_ancestor.pl:58]: Why aren't these changes in the first commit?

**Comment by KnightMurloc** [comment]
> Use checkpoint+1 to account for the edge case when the checkpoint exactly
matches switchpoint LSN.

Is this not the same case as described in the first commit?

Add the appropriate description in the second commit.

**Comment by KnightMurloc** [comment]
If switches to the first commit, the tests pass with unchanged to pg_rewind.

File: src/bin/pg_rewind/pg_rewind.c (hunk 1 above)
Comment (silent-observer) [reply src/bin/pg_rewind/pg_rewind.c]: I think we shouldn't make potentially breaking changes to the target when dry run is enabled. Added explicit note to the description

File: src/bin/pg_rewind/t/010_target_is_ancestor.pl (hunk 2 above)
Comment (silent-observer) [reply src/bin/pg_rewind/t/010_target_is_ancestor.pl:58]: Moved most test changes to the first commit.

**Comment by silent-observer** [comment]
I split the patch into three commits now.
- The first contains the assert removal and an updated test that can be adapted to second and third commits with minimal changes.
- The second contains the bulk of the patch by Heikki, with minor fixes, and updates to the test that only pass after the second commit.
- The third commit contains a fix for minRecoveryPoint edge case, and extends the test more to cover it.
Now the test changes are fully sequential: the tests pass on each of the three commits, but only with corresponding changes to pg_rewind present.

File: src/bin/pg_rewind/pg_rewind.c [hunk 3]
```diff
@@ -372,18 +375,38 @@ main(int argc, char **argv)
...
 		{
-			/* the last common checkpoint record must be part of target WAL */
-			Assert(target_wal_endrec == divergerec);
+			ControlFileData ControlFile_new;
 
 			rewind_needed = false;
+
+			memcpy(&ControlFile_new, &ControlFile_target, sizeof(ControlFileData));
+			ControlFile_new.minRecoveryPoint = 	Max(ControlFile_source.minRecoveryPoint,
+													ControlFile_source.checkPoint + 1);
```
Comment (KnightMurloc) [inline src/bin/pg_rewind/pg_rewind.c]: As I understand it, LSN should point to the beginning of the XLog record. If we just do +1, then LSN will point inside some record. How normal is this?

Comment (KnightMurloc) [inline src/bin/pg_rewind/pg_rewind.c]: Do we really need to update minRecoveryPoint? Is it not enough to update minRecoveryPointTLI?

File: src/bin/pg_rewind/pg_rewind.c [hunk 4]
```diff
@@ -372,18 +375,38 @@ main(int argc, char **argv)
...
 		 */
 		if (target_wal_endrec > divergerec)
-		{
 			rewind_needed = true;
-		}
 		else
 		{
-			/* the last common checkpoint record must be part of target WAL */
-			Assert(target_wal_endrec == divergerec);
+			ControlFileData ControlFile_new;
```
Comment (KnightMurloc) [inline src/bin/pg_rewind/pg_rewind.c]: You're shadowing a variable. line 133

File: src/bin/pg_rewind/pg_rewind.c [hunk 5]
```diff
@@ -660,7 +683,11 @@ getTimelineHistory(ControlFileData *controlFile, int *nentries)
 
 		/* Get history file from appropriate source */
 		if (controlFile == &ControlFile_source)
+		{
 			histfile = fetchFile(path, NULL);
+			sourceHistfileName = pstrdup(path);
+			sourceHistfile = pstrdup(histfile);
```
Comment (KnightMurloc) [inline src/bin/pg_rewind/pg_rewind.c]: I don't think it's a good idea to modify the global state from this function. Is there any way to put this logic into a separate function to get the contents of the history file?

File: src/bin/pg_rewind/pg_rewind.c (hunk 3 above)
Comment (silent-observer) [reply src/bin/pg_rewind/pg_rewind.c]: `minRecoveryPoint` actually specifies the least acceptable end of WAL. If it points inside some record, that simply guarantees that the record will be processed. This is definitely okay for GPDB since `minRecoveryPoint` isn't used in that many places, and should probably be okay for raw Postgres too, although that'd require more research and discussion with upstream.

Comment (silent-observer) [reply src/bin/pg_rewind/pg_rewind.c]: At startup there are checks that `minRecoveryPoint` is on the correct TLI, so no, we have to update them both. It's technically possible to not update the control file at all, but it's less robust and might fail in unexpected ways

File: src/bin/pg_rewind/pg_rewind.c (hunk 4 above)
Comment (silent-observer) [reply src/bin/pg_rewind/pg_rewind.c]: Fixed

File: src/bin/pg_rewind/pg_rewind.c (hunk 5 above)
Comment (silent-observer) [reply src/bin/pg_rewind/pg_rewind.c]: Unfortunately putting it into a separate function complicates the code a lot since the history file is also immediately processed in this function. I think it's easier to just pass them through out parameters instead.

File: src/bin/pg_rewind/t/107_empty_conf.pl [hunk 6]
```diff
@@ -31,6 +31,10 @@ sub run_test
 
 	RewindTest::run_pg_rewind($test_mode, do_not_start_primary => 1);
 
+	# Append new file to the old one to preserve primary_conninfo
+	my $new_file_data = slurp_file("$master_pgdata/postgresql.conf");
+	append_to_file("$tmp_folder/master-postgresql-full.conf.tmp", $new_file_data);
```
Comment (KnightMurloc) [inline src/bin/pg_rewind/t/107_empty_conf.pl:36]: Why was this fix required?

File: src/bin/pg_rewind/pg_rewind.c (hunk 3 above)
Comment (KnightMurloc) [reply src/bin/pg_rewind/pg_rewind.c]: Can you add some comment about this? Also, why can't we put `InvalidXLogRecPtr` here?

Comment (silent-observer) [reply src/bin/pg_rewind/pg_rewind.c]: `InvalidXLogRecPtr` would disable `minRecoveryPointTLI` as well, so it won't be guaranteed anymore that the standby will rewind to the switchpoint. If we want to use `minRecoveryPoint` functionality at all, we have to specify both of them. I will add the comment

File: src/bin/pg_rewind/t/107_empty_conf.pl (hunk 6 above)
Comment (silent-observer) [reply src/bin/pg_rewind/t/107_empty_conf.pl:36]: pg_rewind adds connection info to the config. Previously because minRecoveryPoint wasn't set and timeline files weren't copied, standby could finish recovery just from the already present WALs. This way it worked even with the original config, not modified by pg_rewind. However, after this patch the standby is now aware about the new timeline on primary and tried to retrieve it during recovery (which failed, because there is no primary_conninfo in the original config). So after this patch, we have to preserve changes made by pg_rewind, and not simply replace the config before starting the standby.

**Review by KnightMurloc — DISMISSED** [review]

**Comment by RekGRpth** [comment]
Is it expected that changes in the `src/bin/pg_rewind/t/107_empty_conf.pl` test work without a patch?

**Comment by RekGRpth** [comment]
> Refactor sourceHistfile to be an out parameter instead of global state.

pgindent these your changes

**Comment by silent-observer** [comment]
> Is it expected that changes in the src/bin/pg_rewind/t/107_empty_conf.pl test work without a patch?

Yes, changes to 107_empty_conf.pl only fix config line being accidentally deleted, they don't change the nature of the test. 010_target_is_ancestor.pl is the main test for this patch.

**Review by RekGRpth — APPROVED** [review]

**Review by KnightMurloc — APPROVED** [review]
